Recovery E2E (#315)
Azure-Samples/iot-middleware-freertos-samples: demos/projects/ESPRESSIF/az-ca-recovery/README.md

# Perform Trust Bundle Recovery with an ESPRESSIF ESP32 using Azure IoT Middleware for FreeRTOS

> Trust Bundle Recovery allows a mechanism for your device to recover connection to Azure resources should the CA certificates expire or otherwise become invalid. This requires an additional, separate "recovery" DPS instance and device credential, only to be used for recovery of the CA certificate. Should CA verification fail, the device will connect to the recovery DPS instance, ignoring TLS CA cert validation, and will be returned a signed trust bundle recovery payload from DPS. This payload contains the complete and updated collection of CA certificates, asymmetrically signed by a root key created by the user. The device receives this payload, verifies the authenticity by decrypting and comparing the signature using the saved public key, and then installs the certificates into the device's non-volatile storage (NVS). This permanent storage allows the new certificates to be loaded again should the device restart, removing the need for the device to use the recovery endpoint again. From that point on, the device can connect to the usual DPS endpoint and provisioned hub with full CA trust validation.

Reviewer: I have a mermaid graph that you could add here to better describe the flow (may need some modifications to include the particularities used in the implementation for crypto, payload formats, etc).

Reviewer: ```mermaid
sequenceDiagram
    autonumber
    participant Device
    participant DPS_operational
    participant DPS_recovery
    participant CustomAllocation

    Note right of Device: Device firmware contains:CA TrustBundle &  version<br>RecoveryKey, Device Recovery Identity and IDScope

    loop retry count
        Device-->>DPS_operational: Connect
        DPS_operational-->>Device: Invalid CA Certificate
    end

    Note left of Device: Initiate Recovery
    Device-->>DPS_recovery: Connect using Device Recovery Identity and IDScope
    Note right of Device: !! Ignore CA validation !!
    Note right of Device: Custom Payload: Current TrustBundle version

    DPS_recovery-->>CustomAllocation: Recovery Identity, CA TrustBundle version
    CustomAllocation-->>CustomAllocation: Verify device CA TrustBundle version
    CustomAllocation-->>DPS_recovery: Custom Payload
    Note left of CustomAllocation: Payload: <br> CA TrustBundle, Device Recovery Identity,<br> version, expiryTime <br> Sign(data=payload, key=RecoveryKey)
    DPS_recovery-->>Device: Registration Information (including Custom Payload)
    Device-->>Device: Verify Custom Payload Signature,<br> version and expiryTime
    Device-->>Device: Install CA TrustBundle
    Note left of Device: End Recovery
    Device-->>DPS_operational: Connect
    DPS_operational-->>Device: Valid CA Certificate
```

> Trust Bundle Recovery allows a mechanism for your device to recover connection to Azure resources should the CA certificates expire or otherwise become invalid. This requires a standard DPS instance and connected hub with associated credentials, and then a separate "recovery" DPS instance and credential, only to be used once for recovery. Should CA verification fail, the device will connect to the recovery DPS instance, ignoring CA cert validation, and will be returned a signed trust bundle recovery payload. This payload contains the complete and updated collection of CA certificates, asymmetrically signed by a root key created by the user. The device receives this payload, verifies the authenticity by decrypting and comparing the signature using the saved public key, and then installs the certificates into the device's non-volatile storage (NVS). This permanent storage allows the certificates to be loaded again should the device restart, removing the need for the device to use the recovery endpoint again. From that point on, the device can connect to the usual DPS endpoint and provisioned hub with full CA validation.

Reviewer 2: ```suggestion
Trust Bundle Recovery allows a mechanism for your device to recover connection to Azure resources should the CA certificates expire or otherwise become invalid. This requires an additional separate "recovery" DPS instance and device credential, only to be used for recovery of the CA certificate. Should CA verification fail, the device will connect to the recovery DPS instance, ignoring TLS cert validation, and will be returned a signed trust bundle recovery payload from DPS. This payload contains the complete and updated collection of CA certificates, asymmetrically signed by a root key created by the user. The device receives this payload, verifies the authenticity by decrypting and comparing the signature using the saved public key, and then installs the certificates into the device's non-volatile storage (NVS). This permanent storage allows the new certificates to be loaded again should the device restart, removing the need for the device to use the recovery endpoint again. From that point on, the device can connect to the usual DPS endpoint and provisioned hub with full CA trust validation.
```

- Windows: <https://docs.microsoft.com/windows/win32/fileio/maximum-file-path-limitation?tabs=cmd#enable-long-paths-in-windows-10-version-1607-and-later>
- Git: as Administrator run `git config --system core.longpaths true`

> ## Prepare the Trust Bundle in the NVS

Reviewer 2: ```suggestion
## Prepare the TLS CA Trust Bundle in the NVS
```

> Run the sample called `az-nvs-cert-bundle` in the `/ESPRESSIF` directory to load the v1.0 trust bundle in your ESP device. This will purposely save an incomplete trust bundle in your devices's NVS, which will then be loaded for the IoT application. Once the CA validation fails in this sample, the device will then move into the recovery phase which fetches the new and complete trust bundle.

Reviewer 2: ```suggestion
Run the sample called `az-nvs-cert-bundle` in the `/ESPRESSIF` directory to load the v1.0 trust bundle in your ESP device. This will purposely save an incomplete trust bundle in your devices's NVS, which will then be loaded for the IoT application. Once the CA validation fails in this sample, the device will then move into the recovery phase which fetches the new and complete certificate trust bundle.
```

- Windows: <https://docs.microsoft.com/windows/win32/fileio/maximum-file-path-limitation?tabs=cmd#enable-long-paths-in-windows-10-version-1607-and-later>
- Git: as Administrator run `git config --system core.longpaths true`

## Prepare the TLS CA Trust Bundle in the NVS

> Run the sample called `az-nvs-cert-bundle` in the `/ESPRESSIF` directory to load the v1.0 trust bundle in your ESP device. This will purposely save an incomplete trust bundle in your devices's NVS, which will then be loaded for the IoT application. Once the CA validation fails in this sample, the device will then move into the recovery phase which fetches the new and complete certificate trust bundle.

Reviewer: Instead of going this route, we should document that:
1. The first bundle is Baltimore only
2. Initially Hub is configured with Baltimore (default)
3. Customers will need to use https://learn.microsoft.com/en-us/azure/iot-hub/migrate-tls-certificate?tabs=portal to switch to Digicert
4. Devices will be unable to connect after the change (may take hours due to DNS replication)
5. Device will perform recovery, download the new TrustBundle and successfully connect to Azure IoT.

Author: I worry that complicates the process. This is supposed to show the flow if the certs are invalid and it does that. There's already a lot of steps to flash the device and then there's the whole flow to create the Azure Function, set up a separate DPS and IoT Hub, etc. I would like to keep it more concise to show how this might work.

Reviewer: I take it back: I agree with you that it's likely too late to use my proposal for testing. 
I assume we will be required to update this method once the service starts deploying the new CAs.

 `WiFi SSID` | _{Your WiFi SSID}_
 `WiFi Password` | _{Your WiFi password}_

Under menu item `Azure IoT middleware for FreeRTOS Main Task Configuration`, update the following configuration values:

Parameter | Value
---------|----------
 `Enable Device Provisioning Sample` | _{Check this option to enable DPS in the sample}_
 `Azure Device Provisioning Service ID Scope` | _{Your ID scope value}_
 `Azure Device Provisioning Service Registration ID` | _{Your Device Registration ID value}_
 `[RECOVERY] Azure Device Provisioning Service ID Scope` | _{Your ID scope value for the recovery instance}_
 `[RECOVERY] Azure Device Provisioning Service Registration ID.` | _{Your Device Registration ID value for the recovery instance}_

> Select your desired authentication method with the `Azure IoT Authentication Method () --->` as `Symmetric Key`:

Reviewer: Do we support all auth systems with recovery and operational DPS / Hub services in this sample?
(If not, we should update these to only what is supported.)

Author: No this is symmetric key only right now.

Parameter | Value
---------|----------
 `Azure IoT Device Symmetric Key` | _{Your Azure IoT Hub device symmetric key}_
 `[RECOVERY] Azure IoT Device Symmetric Key` | _{Your Azure IoT Hub device symmetric key for the recovery instance}_

Save the configuration (`Shift + S`) inside the sample folder in a file with name `sdkconfig`.
After that, close the configuration utility (`Shift + Q`).

You must also update the signing root key which is located in `demo_config.h`, titled `ucAzureIoTRecoveryRootKeyN`. You can get the hex value of the modulus (N value) using the below command.

```bash

> openssl x509 -in <your-public-cert>.pem -modulus -noout | sed s/Modulus=// | sed -r 's/../0x&, /g'

Reviewer: Do we use a default exponent of 65537? 
AFAIK RSA will require both N and E.

Author: Yes we default to 65537. I can make a note we set it to that by default (and most do), but they might need to update it.
